Fixed #19721 -- Allowed admin filters to customize the list separator.

ssb10 opened the pull request:
References ticket https://code.djangoproject.com/ticket/19721
The ticket reports a bug that occurs when admin field lookup using `__in` contains a comma in the value.
Example: Looking up for string `Jack,Red`
`my/request/name__in=Jack,Red` returns an empty queryset since the search value is split on ',' for `__in` lookups.

I referred to the [patch ](https://code.djangoproject.com/attachment/ticket/19721/comma_in__in_data.patch)attached to the ticket.

carltongibson reviewed (commented):
Hi @ssb10, thanks for this! 
 
Alas, I think there's a bit more to wrapping up this ticket than applying the old patch. 

Looking at the suggested regex, I **really** 👀 don't think it's a good trade-off of benefit to complexity. (That this ticket has sat for 7 years suggests the demand is not high.) 

[Carl Meyer’s suggestion on the ticket](https://code.djangoproject.com/ticket/19721#comment:3) suggested the way forward I think: 

> …need to provide a way to easily subclass the standard ListFilter and replace just the separator character (and then document that). 

This should be significantly simpler. 

1. Add a `divider=','` parameter to `prepare_lookup_value()`. 
2. Pass that from an attribute in `FieldListFilter` when calling in `__init__()`. 
3. Document with an example showing a `FieldListFilter` subclass setting the new attribute. 

It's step 3 that's most work. The [current `list_filter` docs](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.list_filter) are somewhat overloaded, so maybe we need to consider a small restructuring. (We can deal with that once a draft is in place, however.) 

Does that all make sense? Let me know. 

Thanks again!

ssb10 commented:
> This should be significantly simpler.
> 1. Add a divider=',' parameter to prepare_lookup_value().
> 2. Pass that from an attribute in FieldListFilter when calling in __init__().
> 3. Document with an example showing a FieldListFilter subclass setting the new attribute.

@carltongibson Thank you for your help! I made updates as suggested in steps 1 and 2. I hope I got the logic right.
If the new changes look right, I can update the documentation and I'll be happy to do the restructuring of docs with some input from you. :)

carltongibson reviewed (commented):
Hi @ssb10 — thanks for the update. 

You've added `divider` being passed from the query string... — no real problem with that — it's a bit more API than I thought initially: I would have made folks subclass to opt into this — so they'd fall into the `This is simply a custom list filter class.` block — but leave it for the moment. 🤔 

For the docs, I'd be inclined to add a `filters.txt` in `docs/ref/contrib/admin`. (`index.html` is already _quite long_…) If we could shrink the `.. attribute:: ModelAdmin.list_filter` section and then have a `See <new page> for more details`, I think that would be a win. 

I guess I'd copy out what's there and adjust to begin. 

Are you happy to begin a draft? (It doesn't have to be perfect first go. We can discuss.) 

Thanks for your effort on this! 🎁

ssb10 commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -0,0 +1,162 @@
+
+.. currentmodule:: django.contrib.admin
+
+.. attribute:: ModelAdmin.list_filter :noindex:
```
Added `:no-index:` here since Sphinx was showing a warning for two attributes with same names. 
Should it be attribute though in this file?

ssb10 commented:
@carltongibson I can try to implement the subclass approach you mentioned above.
I have started restructuring the docs. Created a new file for list filters and added a link in the index.txt file. I am trying to preview the changes in Sphinx but I have not been able to do so. I think I'll figure out how to preview stuff so that I can verify my changes look right.
Thank you for your help!

ssb10 commented:
Update: I was able to preview the documentation. Attaching the screenshot for reference
1. index.html
<img width="549" alt="list_filter_index_html" src="https://user-images.githubusercontent.com/29463692/141372527-ad43c0df-1385-4b07-b4e5-dbe8348f6e11.PNG">

2. filters.html
<img width="738" alt="list_filter_html" src="https://user-images.githubusercontent.com/29463692/141372599-28b0ca1f-6f09-4baa-9ea8-1162da6cf688.PNG">
.

ssb10 commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -0,0 +1,163 @@
+
+.. currentmodule:: django.contrib.admin
+
+.. attribute:: ModelAdmin.list_filter
+    :noindex:
```
Added :no-index: here since Sphinx was showing a warning for two attributes with same names.

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -852,160 +853,8 @@ subclass::
-
-    See the default template provided by Django (``admin/filter.html``) for
-    a concrete example.
+    See the :doc:`list_filter documentation</ref/contrib/admin/filters>`
+    for more details.
```
We should leave some stub here, just a sentence, saying what `list_filter` is and then linking. 

If you add a hyperlink target to the `filters.txt` you can link with `:ref:`admin-list-filters`, which is more robust over time.

carltongibson commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -0,0 +1,162 @@
+=================
```
You can add a target here to link to: 

```suggestion
.. _admin-list-filters:

=================
```

carltongibson commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -0,0 +1,162 @@
+
+.. currentmodule:: django.contrib.admin
+
+.. attribute:: ModelAdmin.list_filter :noindex:
```
I think this `attribute` declaration (still) belongs back in the reference. 

This doc is providing details on the allowed values for ``` :ref:`ModelAdmin.list_filter` ```
It's a free standing doc, so it probably wants a introductory paragraph saying such. 

Then we can outdent the list items below. The only reason they're presented as a list, I think, is that they had to fit in the context of the larger doc. Now they're independent, it can be restructured. 

How about an overview stating the options (as a list say) and then sections for each possibility, which allows us to expand with the examples we'll need for this PR. 

I think moving to it's own page, plus small restructuring like this, could be merged as a single commit before adjusting the content for the new functionality.

ssb10 replied:
- Removed the `attribute` declaration from filters.txt
- Added an overview stating the options for allowed values for `ModelAdmin.list_filter` in a numbered list.
- I have removed the details of the examples related to this PR and will add them after these changes are finalized.

carltongibson reviewed (commented):
Hi @ssb10 — thanks for the updates. I added a some comments. 

1. Let's aim for a first commit that just extracts the `list_filter` details (as you've done) and restructures slightly, to make it flatter now that the content is in it's own page. Then we can add the new content. 

     - If you push edits just on that goal I can extract them to a single commit. 
     - Do my comments make sense to you? Are you OK to try those edits? 

2. Well done getting Sphinx going. We can review the issues you had and make a change to the contributing guide in a separate PR. (It can easily be that the Windows version of the instructions are not as correct, since they're less often travelled.)

ssb10 commented:
> 1. Let's aim for a first commit that just extracts the `list_filter` details (as you've done) and restructures slightly, to make it flatter now that the content is in it's own page. Then we can add the new content.
>    
>    * If you push edits just on that goal I can extract them to a single commit.
>    * Do my comments make sense to you? Are you OK to try those edits?

I have made an attempt to restructure the `ModelAdmin.list_filter` docs. I am happy to make edits to the doc (I know it is not perfect. I referred to existing documentation and attempted to restructure the docs).
Yes I have understood the comments and made changes accordingly. Let me know if I missed something or misunderstood the comment.

> 2. Well done getting Sphinx going. We can review the issues you had and make a change to the contributing guide in a separate PR. (It can easily be that the Windows version of the instructions are not as correct, since they're less often travelled.)

Thank you! I hope it helps the Windows users in the community.

kezabelle commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -0,0 +1,181 @@
+                ('title', admin.EmptyFieldListFilter),
+            )
+
+.. note::
+
+    The ``FieldListFilter`` API is considered internal and might be
+    changed.
```
Just to get @carltongibson's eyes on it; I know it's just moving what was previously there:

This line has been in place since 2012, despite there being prior examples of using a subclass (e.g. `BooleanFieldListFilter`) previously.
The API itself might well be considered internal, but it's been promoted in the docs here for nearly a decade, so people will be using it, so it would probably be subject to the usual  deprecation path on API change anyway. Also the `FieldListFilter` has only been touched once since 2012, to fix [ticket 28202](https://code.djangoproject.com/ticket/28202).

What I'm saying is either as part of this PR, or a subsequent one, this scary warning could probably be considered for removal?

carltongibson replied:
Thanks @kezabelle — Yes. Going over all this now, we should probably document this fully, with the available subclasses too.

carltongibson replied:
@ssb10 A third possible PR there if you want it once we've got this in. 😄 

(Windows/Sphinx set-up review being the second one.)

ssb10 replied:
Yes sure. I will definitely give it a shot. 🙂

carltongibson commented:
Hey @ssb10 — I've pushed a commit with the first docs changes in #15097. Can you take a look and comment? (Thanks) 

The idea would be to merge that change and then rebase this after that to add the new behaviour (and docs). Make sense?

ssb10 commented:
Hey @carltongibson, I just commented on the docs PR. Everything looks good to me. Thank you for the edits 🙌 

Yes once the other PR is merged I can rebase my branch on main and add the docs for new behaviour.

carltongibson commented:
Hi @ssb10 — e53aea2e23b592c4774f0e9fa7f9f9d726a40dfc is in — Thanks for your work on that! 

Can you rebase the changes here after that? It looks like dropping the changes in the `doc` files will allow the rest of the changes here to apply. 

(Let me know if you need input on rebasing. Git is… well… _particular_, and I don't want it to be a barrier.)

ssb10 commented:
@carltongibson thank you for your help with the `ModelAdmin.list_filter` documentation. 
I have pulled the new changes and rebased this branch and also updated filters.txt with the new documentation.

One small thing to note: I built the docs after adding my changes and I see some incorrect formatting on the filters.html page. (See screenshot). Is it possible that the third bullet, "A 2-tuple containing ..." looks the way it does only in this preview or it needs fixing?
<img width="729" alt="list_filter_new" src="https://user-images.githubusercontent.com/29463692/142486934-4055853e-5ca3-4d81-bc8a-ee8a511c8b42.PNG">

carltongibson commented:
@ssb10 Good spot. Not sure. Will look 🧐. 

If you fancied, you could squash your commits down and use the [project commit message format](https://docs.djangoproject.com/en/3.2/internals/contributing/committing-code/#committing-guidelines). (No rush on that.)

felixxm commented:
> @ssb10 Good spot. Not sure. Will look monocle_face.

It's probably caused by an extra space:
```diff
diff --git a/docs/ref/contrib/admin/filters.txt b/docs/ref/contrib/admin/filters.txt
index a0cd931ce4..f78005936d 100644
--- a/docs/ref/contrib/admin/filters.txt
+++ b/docs/ref/contrib/admin/filters.txt
@@ -18,7 +18,7 @@ or tuple of elements, where each element is one of the following types:
 - A field name.
 - A subclass of ``django.contrib.admin.SimpleListFilter``.
 - A 2-tuple containing a field name and a subclass of
-   ``django.contrib.admin.FieldListFilter``.
+  ``django.contrib.admin.FieldListFilter``.
 
 See the examples below for discussion of each of these options for defining
 ``list_filter``.
```

kezabelle commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -176,6 +176,18 @@ allows to store::
         )
 
+It is also possible to specify a custom divider for list filters
+while using the ``__in`` lookup. By default ``__in`` lookup uses comma
```
> By default `__in` lookup uses comma

maybe should be

> By default _the_ `__in` lookup uses _a_ comma

emphasis for clarity.

kezabelle commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -176,6 +176,18 @@ allows to store::
+    class FilterWithCustomDivider(admin.CustomDividerFilter):
+        def get_divider(self):
+            # return the custom divider as a string.
+            return '%'
```
Just a thought - perhaps consider a different example character; % is already used for URL encoding (so `%` _can_ also be `%25`). It's probably not an _issue_ per se, mind.

ssb10 replied:
I agree. I will use a different character.

ssb10 commented:
> If you fancied, you could squash your commits down and use the [project commit message format](https://docs.djangoproject.com/en/3.2/internals/contributing/committing-code/#committing-guidelines). (No rush on that.)

@carltongibson sure I can squash commits and use the project commit message format. I have a question about the same. Is it okay to drop the earlier commits which are not at all related to the latest ones? For example, the initial commits where I implemented logic to escape commas?

carltongibson commented:
@ssb10 Yes, it's fine to drop outdated/exploratory work. What we're after (ultimately) is a single commit that contains just the changes needed to fix the ticket. :)

tim-mccurrach commented:
Thanks for your work on this patch :) This looks good 👍 I have a few thoughts:

1) Would `separator` be a better name than `divider` in most cases here? It would fit with what seems to be the standard naming convention among built-in functions in python (and django). That is `print`, `django.utils.numberformat`, and various other python built-ins all use `separator` (or some more specific variant) for this kind of thing.

2) Is a full-blown method needed to specify the separator (divider)? I can't imagine a case where it would be different with multiple calls. Could we make it a class attribute? Say `list_separator`.

3) And if we're doing that, couldn't we just make it a class attribute on `FieldListFilter` rather than a new special class. Then all that would be needed would be for users to subclass `FieldListFilter`. I appreciate users would then need to handle `expected_parameters` and `lookup_kwarg` themselves, but anybody using `__in` will be doing that already anyway. And this way, we keep things more flexible, users can do `x__y__in` for example. Further, if someone subclasses some other `FieldListFilter`-subclass and overwrites `expected_parameters` to return something ending in `__in` then they'll potentially need this functionality there as well. 

I appreciate that a fair amount of thought and effort has already gone into this PR, and I'm aware that I'm joining the party quite late in the day - so I hope I'm not muddying the waters too much here :/

ssb10 commented:
> @ssb10 Yes, it's fine to drop outdated/exploratory work. What we're after (ultimately) is a single commit that contains just the changes needed to fix the ticket. :)

Thank you @carltongibson for the clarification. I have squashed/dropped commits. 
Also updated the logic as per @tim-mccurrach's suggestions (also updated the docs). If the new changes look good, I can also cleanup more by removing custom divider filter commits too. :)

carltongibson commented:
Hi @ssb10 — sorry for the delay on this — I was on holiday, and then moving house. Picking up for final review now. 👍

carltongibson reviewed (approved):
Hi @ssb10 — thanks for this, really good. 

I just made a few edits, added a release note and rebased. ⛵

carltongibson commented:
Aside: I think the filter docs here are a little sparse still — confusing even — I'm not sure I'd work it out without opening the source and quite a bit of trial and error. 

I think the change in this PR is good, and reasonably scoped, but overall a bit more restructuring/work/love wouldn't go amiss in this direction (for whoever wants to pick that up at any time.) 

Thanks again @ssb10!

ssb10 commented:
Thank you @carltongibson for the review and your help throughout this PR.
I was wondering if I can open a separate PR for the windows-sphinx documentation - 

>Well done getting Sphinx going. We can review the issues you had and make a change to the contributing guide in a separate PR. (It can easily be that the Windows version of the instructions are not as correct, since they're less often travelled.)

(Also I was wondering if there is a better place to ask questions like this than Github comments. May be on IRC?)

carltongibson commented:
Hi @ssb10. 

> I was wondering if I can open a separate PR for the windows-sphinx documentation

Yes, please. (If it's just a small change no ticket needed. If it's bigger then yes...) 🏅

> Also I was wondering if there is a better place to ask questions like this than Github comments. May be on IRC?

IRC works. `#django-dev` channel. (There's always someone in there, but I'm not very good at it myself — you can try me though :) 

Also, the [Forum has a Mentoring Channel](https://forum.djangoproject.com/c/internals/mentorship/10). You can @carltongibson me there and I'll definitely see that (though others will help you too.) 

I know you've made a few commits now but, Welcome aboard! ⛵
